// @stacks/stacking/dist/esm/index.js
import { Buffer, intToBigInt } from '@stacks/common';
import { makeContractCall, bufferCV, uintCV, tupleCV, ClarityType, broadcastTransaction, standardPrincipalCV, callReadOnlyFunction, cvToString, noneCV, someCV, validateStacksAddress, AnchorMode, } from '@stacks/transactions';
import BN from 'bn.js';
import { StackingErrors } from './constants';
import { fetchPrivate } from '@stacks/common';
import { decodeBtcAddress } from './utils';
export * from './utils';
export class StackingClient {
    constructor(address, network) {
        this.address = address;
        this.network = network;
    }
    async getCoreInfo() {
        const url = this.network.getInfoUrl();
        return fetchPrivate(url).then(res => res.json());
    }
    async getPoxInfo() {
        const url = this.network.getPoxInfoUrl();
        return fetchPrivate(url).then(res => res.json());
    }
    async getTargetBlockTime() {
        const url = this.network.getBlockTimeInfoUrl();
        const res = await fetchPrivate(url).then(res => res.json());
        if (this.network.isMainnet()) {
            return res.mainnet.target_block_time;
        }
        else {
            return res.testnet.target_block_time;
        }
    }
    async getAccountStatus() {
        const url = this.network.getAccountApiUrl(this.address);
        return fetchPrivate(url).then(res => res.json());
    }
    async getAccountBalance() {
        return this.getAccountStatus().then(res => {
            return BigInt(res.balance);
        });
    }
    async getCycleDuration() {
        const poxInfoPromise = this.getPoxInfo();
        const targetBlockTimePromise = await this.getTargetBlockTime();
        return Promise.all([poxInfoPromise, targetBlockTimePromise]).then(([poxInfo, targetBlockTime]) => {
            return poxInfo.reward_cycle_length * targetBlockTime;
        });
    }
    async getRewardsTotalForBtcAddress() {
        const url = this.network.getRewardsTotalUrl(this.address);
        return fetchPrivate(url).then(res => res.json());
    }
    async getRewardsForBtcAddress(options) {
        const url = `${this.network.getRewardsUrl(this.address, options)}`;
        return fetchPrivate(url).then(res => res.json());
    }
    async getRewardHoldersForBtcAddress(options) {
        const url = `${this.network.getRewardHoldersUrl(this.address, options)}`;
        return fetchPrivate(url).then(res => res.json());
    }
    async getSecondsUntilNextCycle() {
        const poxInfoPromise = this.getPoxInfo();
        const targetBlockTimePromise = await this.getTargetBlockTime();
        const coreInfoPromise = this.getCoreInfo();
        return Promise.all([poxInfoPromise, targetBlockTimePromise, coreInfoPromise]).then(([poxInfo, targetBlockTime, coreInfo]) => {
            const blocksToNextCycle = poxInfo.reward_cycle_length -
                ((coreInfo.burn_block_height - poxInfo.first_burnchain_block_height) %
                    poxInfo.reward_cycle_length);
            return blocksToNextCycle * targetBlockTime;
        });
    }
    async isStackingEnabledNextCycle() {
        return (await this.getPoxInfo()).rejection_votes_left_required > 0;
    }
    async hasMinimumStx() {
        const balance = await this.getAccountBalance();
        const min = BigInt((await this.getPoxInfo()).min_amount_ustx);
        return balance >= min;
    }
    async canStack({ poxAddress, cycles }) {
        const balancePromise = this.getAccountBalance();
        const poxInfoPromise = this.getPoxInfo();
        return Promise.all([balancePromise, poxInfoPromise])
            .then(([balance, poxInfo]) => {
            const { hashMode, data } = decodeBtcAddress(poxAddress);
            const hashModeBuffer = bufferCV(new BN(hashMode, 10).toArrayLike(Buffer));
            const hashbytes = bufferCV(data);
            const poxAddressCV = tupleCV({
                hashbytes,
                version: hashModeBuffer,
            });
            const [contractAddress, contractName] = this.parseContractId(poxInfo.contract_id);
            return callReadOnlyFunction({
                network: this.network,
                contractName,
                contractAddress,
                functionName: 'can-stack-stx',
                senderAddress: this.address,
                functionArgs: [
                    poxAddressCV,
                    uintCV(balance.toString()),
                    uintCV(poxInfo.reward_cycle_id),
                    uintCV(cycles.toString()),
                ],
            });
        })
            .then((responseCV) => {
            if (responseCV.type === ClarityType.ResponseOk) {
                return {
                    eligible: true,
                };
            }
            else {
                const errorCV = responseCV;
                return {
                    eligible: false,
                    reason: StackingErrors[+cvToString(errorCV.value)],
                };
            }
        });
    }
    async stack({ amountMicroStx, poxAddress, cycles, privateKey, burnBlockHeight, }) {
        const poxInfo = await this.getPoxInfo();
        const contract = poxInfo.contract_id;
        const txOptions = this.getStackOptions({
            amountMicroStx,
            cycles,
            poxAddress,
            contract,
            burnBlockHeight,
        });
        const tx = await makeContractCall(Object.assign(Object.assign({}, txOptions), { senderKey: privateKey }));
        return broadcastTransaction(tx, txOptions.network);
    }
    async delegateStx({ amountMicroStx, delegateTo, untilBurnBlockHeight, poxAddress, privateKey, }) {
        const poxInfo = await this.getPoxInfo();
        const contract = poxInfo.contract_id;
        const txOptions = this.getDelegateOptions({
            contract,
            amountMicroStx,
            delegateTo,
            untilBurnBlockHeight,
            poxAddress,
        });
        const tx = await makeContractCall(Object.assign(Object.assign({}, txOptions), { senderKey: privateKey }));
        return broadcastTransaction(tx, txOptions.network);
    }
    async delegateStackStx({ stacker, amountMicroStx, poxAddress, burnBlockHeight, cycles, privateKey, nonce, }) {
        const poxInfo = await this.getPoxInfo();
        const contract = poxInfo.contract_id;
        const txOptions = this.getDelegateStackOptions({
            contract,
            stacker,
            amountMicroStx,
            poxAddress,
            burnBlockHeight,
            cycles,
            nonce,
        });
        const tx = await makeContractCall(Object.assign(Object.assign({}, txOptions), { senderKey: privateKey }));
        return broadcastTransaction(tx, txOptions.network);
    }
    async stackAggregationCommit({ poxAddress, rewardCycle, privateKey, }) {
        const poxInfo = await this.getPoxInfo();
        const contract = poxInfo.contract_id;
        const txOptions = this.getStackAggregationCommitOptions({
            contract,
            poxAddress,
            rewardCycle,
        });
        const tx = await makeContractCall(Object.assign(Object.assign({}, txOptions), { senderKey: privateKey }));
        return broadcastTransaction(tx, txOptions.network);
    }
    async revokeDelegateStx(privateKey) {
        const poxInfo = await this.getPoxInfo();
        const contract = poxInfo.contract_id;
        const txOptions = this.getRevokeDelegateStxOptions(contract);
        const tx = await makeContractCall(Object.assign(Object.assign({}, txOptions), { senderKey: privateKey }));
        return broadcastTransaction(tx, txOptions.network);
    }
    getStackOptions({ amountMicroStx, poxAddress, cycles, contract, burnBlockHeight, }) {
        const { hashMode, data } = decodeBtcAddress(poxAddress);
        const hashModeBuffer = bufferCV(new BN(hashMode, 10).toArrayLike(Buffer));
        const hashbytes = bufferCV(data);
        const address = tupleCV({
            hashbytes,
            version: hashModeBuffer,
        });
        const [contractAddress, contractName] = this.parseContractId(contract);
        const network = this.network;
        const txOptions = {
            contractAddress,
            contractName,
            functionName: 'stack-stx',
            functionArgs: [uintCV(amountMicroStx), address, uintCV(burnBlockHeight), uintCV(cycles)],
            validateWithAbi: true,
            network,
            anchorMode: AnchorMode.Any,
        };
        return txOptions;
    }
    getDelegateOptions({ contract, amountMicroStx, delegateTo, untilBurnBlockHeight, poxAddress, }) {
        let address = undefined;
        if (poxAddress) {
            const { hashMode, data } = decodeBtcAddress(poxAddress);
            const hashModeBuffer = bufferCV(new BN(hashMode, 10).toArrayLike(Buffer));
            const hashbytes = bufferCV(data);
            address = someCV(tupleCV({
                hashbytes,
                version: hashModeBuffer,
            }));
        }
        const [contractAddress, contractName] = this.parseContractId(contract);
        const network = this.network;
        const txOptions = {
            contractAddress,
            contractName,
            functionName: 'delegate-stx',
            functionArgs: [
                uintCV(amountMicroStx),
                standardPrincipalCV(delegateTo),
                untilBurnBlockHeight ? someCV(uintCV(untilBurnBlockHeight)) : noneCV(),
                address ? address : noneCV(),
            ],
            validateWithAbi: true,
            network,
            anchorMode: AnchorMode.Any,
        };
        return txOptions;
    }
    getDelegateStackOptions({ contract, stacker, amountMicroStx, poxAddress, burnBlockHeight, cycles, nonce, }) {
        const { hashMode, data } = decodeBtcAddress(poxAddress);
        const hashModeBuffer = bufferCV(new BN(hashMode, 10).toArrayLike(Buffer));
        const hashbytes = bufferCV(data);
        const address = tupleCV({
            hashbytes,
            version: hashModeBuffer,
        });
        const [contractAddress, contractName] = this.parseContractId(contract);
        const network = this.network;
        const txOptions = {
            contractAddress,
            contractName,
            functionName: 'delegate-stack-stx',
            functionArgs: [
                standardPrincipalCV(stacker),
                uintCV(amountMicroStx),
                address,
                uintCV(burnBlockHeight),
                uintCV(cycles),
            ],
            validateWithAbi: true,
            network,
            anchorMode: AnchorMode.Any,
        };
        if (nonce) {
            txOptions.nonce = nonce;
        }
        return txOptions;
    }
    getStackAggregationCommitOptions({ contract, poxAddress, rewardCycle, }) {
        const { hashMode, data } = decodeBtcAddress(poxAddress);
        const hashModeBuffer = bufferCV(new BN(hashMode, 10).toArrayLike(Buffer));
        const hashbytes = bufferCV(data);
        const address = tupleCV({
            hashbytes,
            version: hashModeBuffer,
        });
        const [contractAddress, contractName] = this.parseContractId(contract);
        const network = this.network;
        const txOptions = {
            contractAddress,
            contractName,
            functionName: 'stack-aggregation-commit',
            functionArgs: [address, uintCV(rewardCycle)],
            validateWithAbi: true,
            network,
            anchorMode: AnchorMode.Any,
        };
        return txOptions;
    }
    getRevokeDelegateStxOptions(contract) {
        const [contractAddress, contractName] = this.parseContractId(contract);
        const network = this.network;
        const txOptions = {
            contractAddress,
            contractName,
            functionName: 'revoke-delegate-stx',
            functionArgs: [],
            validateWithAbi: true,
            network,
            anchorMode: AnchorMode.Any,
        };
        return txOptions;
    }
    async getStatus() {
        const poxInfo = await this.getPoxInfo();
        const [contractAddress, contractName] = this.parseContractId(poxInfo.contract_id);
        const account = await this.getAccountStatus();
        const functionName = 'get-stacker-info';
        return callReadOnlyFunction({
            contractAddress,
            contractName,
            functionName,
            senderAddress: this.address,
            functionArgs: [standardPrincipalCV(this.address)],
            network: this.network,
        }).then((responseCV) => {
            if (responseCV.type === ClarityType.OptionalSome) {
                const someCV = responseCV;
                const tupleCV = someCV.value;
                const poxAddress = tupleCV.data['pox-addr'];
                const amountMicroStx = tupleCV.data['amount-ustx'];
                const firstRewardCycle = tupleCV.data['first-reward-cycle'];
                const lockPeriod = tupleCV.data['lock-period'];
                const version = poxAddress.data['version'];
                const hashbytes = poxAddress.data['hashbytes'];
                return {
                    stacked: true,
                    details: {
                        amount_microstx: amountMicroStx.value.toString(),
                        first_reward_cycle: Number(firstRewardCycle.value),
                        lock_period: Number(lockPeriod.value),
                        unlock_height: account.unlock_height,
                        pox_address: {
                            version: version.buffer,
                            hashbytes: hashbytes.buffer,
                        },
                    },
                };
            }
            else if (responseCV.type === ClarityType.OptionalNone) {
                return {
                    stacked: false,
                };
            }
            else {
                throw new Error(`Error fetching stacker info`);
            }
        });
    }
    modifyLockTxFee({ tx, amountMicroStx }) {
        const fee = tx.auth.getFee();
        tx.payload.functionArgs[0] = uintCV(intToBigInt(amountMicroStx, false) - fee);
        return tx;
    }
    parseContractId(contract) {
        const parts = contract.split('.');
        if (parts.length !== 2 || !validateStacksAddress(parts[0]) || parts[1] !== 'pox') {
            throw new Error('Stacking contract ID is malformed');
        }
        return parts;
    }
}
